### PR TITLE
[Merged by Bors] - feat: port GroupTheory.Subsemigroup.Operations

### PR DESCRIPTION
**Pull request by ChrisHughes24**

### CODE REVIEW

**Comment by alreadydone** [comment]
Maybe you could check if the changes [here](https://github.com/leanprover-community/mathlib/pull/18015/files#diff-de451a04648ef1a861af801915ac4f8aa3542634d647f873240a461af72dc581) is necessary for Lean 4, and make the corresponding changes in this PR if they are?

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 1]
```diff
@@ -0,0 +1,899 @@
...
+
+variable [Mul M]
+
+/-- Subsemigroups of semigroup `M` are isomorphic to additive subsemigroups of `additive M`. -/
+@[simps]
+def Subsemigroup.toAddSubsemigroup :
+    Subsemigroup M ≃o
+      AddSubsemigroup
+        (Additive
+          M) where
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:88]:
```suggestion
def Subsemigroup.toAddSubsemigroup : Subsemigroup M ≃o AddSubsemigroup (Additive M) where
```

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 2]
```diff
@@ -0,0 +1,899 @@
...
+## Tags
+
+subsemigroup, range, product, map, comap
+-/
+
+
+variable {M N P σ : Type _}
+
+/-!
+### Conversion to/from `additive`/`multiplicative`
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:74]:
```suggestion
### Conversion to/from `Additive`/`Multiplicative`
```

Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:74]:
```suggestion
### Conversion to/from `Additive`/`Multiplicative`
```

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 3]
```diff
@@ -0,0 +1,899 @@
...
+-/
+
+
+section
+
+variable [Mul M]
+
+/-- Subsemigroups of semigroup `M` are isomorphic to additive subsemigroups of `additive M`. -/
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:82]:
```suggestion
/-- Subsemigroups of semigroup `M` are isomorphic to additive subsemigroups of `Additive M`. -/
```

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 4]
```diff
@@ -0,0 +1,899 @@
...
+variable {A : Type _} [Add A]
+
+/-- Additive subsemigroups of an additive semigroup `A` are isomorphic to
+multiplicative subsemigroups of `Multiplicative A`. -/
+@[simps]
+def AddSubsemigroup.toSubsemigroup :
+    AddSubsemigroup A ≃o
+      Subsemigroup
+        (Multiplicative
+          A) where
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:137]:
```suggestion
def AddSubsemigroup.toSubsemigroup : AddSubsemigroup A ≃o Subsemigroup (Multiplicative A) where
```

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 5]
```diff
@@ -0,0 +1,899 @@
...
+theorem comap_id (S : Subsemigroup P) : S.comap (MulHom.id _) = S :=
+  ext (by simp)
+#align subsemigroup.comap_id Subsemigroup.comap_id
+
+/-- The image of a subsemigroup along a semigroup homomorphism is a subsemigroup. -/
+@[to_additive
+      "The image of an `AddSubsemigroup` along an `AddSemigroup` homomorphism is
+      an `AddSubsemigroup`."]
+def map (f : M →ₙ* N) (S : Subsemigroup M) :
+    Subsemigroup N where
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:222]:
```suggestion
def map (f : M →ₙ* N) (S : Subsemigroup M) : Subsemigroup N where
```

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 6]
```diff
@@ -0,0 +1,899 @@
...
+namespace MulMemClass
+
+variable {A : Type _} [Mul M] [SetLike A M] [hA : MulMemClass A M] (S' : A)
+
+-- lower priority so other instances are found first
+/-- A submagma of a magma inherits a multiplication. -/
+@[to_additive "An additive submagma of an additive magma inherits an addition."]
+instance (priority := 900) hasMul : Mul S' :=
+  ⟨fun a b => ⟨a.1 * b.1, mul_mem a.2 b.2⟩⟩
+#align mul_mem_class.has_mul MulMemClass.hasMul
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:484]:
```suggestion
instance (priority := 900) mul : Mul S' :=
  ⟨fun a b => ⟨a.1 * b.1, mul_mem a.2 b.2⟩⟩
#align mul_mem_class.has_mul MulMemClass.mul
```

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 7]
```diff
@@ -0,0 +1,899 @@
...
+
+-- lower priority so later simp lemmas are used first; to appease simp_nf
+@[simp, norm_cast, to_additive]
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:487]: What priority?

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 8]
```diff
@@ -0,0 +1,899 @@
...
+@[simp, to_additive]
+theorem mk_mul_mk (x y : M) (hx : x ∈ S') (hy : y ∈ S') :
+    (⟨x, hx⟩ : S') * ⟨y, hy⟩ = ⟨x * y, mul_mem hx hy⟩ :=
+  rfl
+#align mul_mem_class.mk_mul_mk MulMemClass.mk_mul_mk
+
+@[to_additive]
+theorem mul_def (x y : S') : x * y = ⟨x * y, mul_mem x.2 y.2⟩ :=
+  rfl
+#align mul_mem_class.mul_def MulMemClass.mul_def
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:502]: Should all the additive versions be `#align`ed, btw?

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 9]
```diff
@@ -0,0 +1,899 @@
...
+
+/-- A subsemigroup of a `CommSemigroup` is a `CommSemigroup`. -/
+@[to_additive "An `AddSubsemigroup` of an `AddCommSemigroup` is an `AddCommSemigroup`."]
+instance toCommSemigroup {M} [CommSemigroup M] {A : Type _} [SetLike A M] [MulMemClass A M]
+    (S : A) : CommSemigroup S :=
+  Subtype.coe_injective.commSemigroup Subtype.val fun _ _ => rfl
+#align mul_mem_class.to_comm_semigroup MulMemClass.toCommSemigroup
+
+/-- The natural semigroup hom from a subsemigroup of semigroup `M` to `M`. -/
+@[to_additive "The natural semigroup hom from an `add_subsemigroup` of `add_semigroup` `M` to `M`."]
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:519]:
```suggestion
@[to_additive "The natural semigroup hom from an `AddSubsemigroup` of `AddSubsemigroup` `M` to `M`."]
```

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 10]
```diff
@@ -0,0 +1,899 @@
...
+end MulMemClass
+
+namespace Subsemigroup
+
+variable [Mul M] [Mul N] [Mul P] (S : Subsemigroup M)
+
+/-- The top subsemigroup is isomorphic to the semigroup. -/
+@[to_additive "The top additive subsemigroup is isomorphic to the additive semigroup.", simps]
+def topEquiv : (⊤ : Subsemigroup M) ≃*
+      M where
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:538]:
```suggestion
def topEquiv : (⊤ : Subsemigroup M) ≃* M where
```

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 11]
```diff
@@ -0,0 +1,899 @@
...
+  left_inv x := x.eta _
+  right_inv _ := rfl
+  map_mul' _ _ := rfl
+#align subsemigroup.top_equiv Subsemigroup.topEquiv
+
+@[simp, to_additive]
+theorem top_equiv_toMulHom :
+    (topEquiv : _ ≃* M).toMulHom = MulMemClass.subtype (⊤ : Subsemigroup M) :=
+  rfl
+#align subsemigroup.top_equiv_to_mul_hom Subsemigroup.top_equiv_toMulHom
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:550]:
```suggestion
theorem topEquiv_toMulHom :
    (topEquiv : _ ≃* M).toMulHom = MulMemClass.subtype (⊤ : Subsemigroup M) :=
  rfl
#align subsemigroup.top_equiv_to_mul_hom Subsemigroup.topEquiv_toMulHom
```

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 12]
```diff
@@ -0,0 +1,899 @@
...
+@[simp, to_additive]
+theorem closure_closure_coe_preimage {s : Set M} :
+    closure ((Subtype.val : closure s → M) ⁻¹' s) = ⊤ :=
+  eq_top_iff.2 fun x =>
+    Subtype.recOn x fun x hx _ => by
+      refine' closure_induction' _ (fun g hg => subset_closure hg) (fun g₁ g₂ hg₁ hg₂ => _) hx
+      · exact Subsemigroup.mul_mem _
+#align subsemigroup.closure_closure_coe_preimage Subsemigroup.closure_closure_coe_preimage
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:573]:
```suggestion
```

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 13]
```diff
@@ -0,0 +1,899 @@
...
+@[to_additive Prod
+      "Given `AddSubsemigroup`s `s`, `t` of `AddSemigroup`s `A`, `B` respectively,
+      `s × t` as an `AddSubsemigroup` of `A × B`."]
+def prod (s : Subsemigroup M) (t : Subsemigroup N) :
+    Subsemigroup (M × N) where
+  carrier := s ×ˢ t
+  mul_mem' hp hq := ⟨s.mul_mem hp.1 hq.1, t.mul_mem hp.2 hq.2⟩
+#align subsemigroup.prod Subsemigroup.prod
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:585]:
```suggestion
```

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 14]
```diff
@@ -0,0 +1,899 @@
...
+#align subsemigroup.closure_closure_coe_preimage Subsemigroup.closure_closure_coe_preimage
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
+/-- Given `Subsemigroup`s `s`, `t` of semigroups `M`, `N` respectively, `s × t` as a subsemigroup
+of `M × N`. -/
+@[to_additive Prod
+      "Given `AddSubsemigroup`s `s`, `t` of `AddSemigroup`s `A`, `B` respectively,
+      `s × t` as an `AddSubsemigroup` of `A × B`."]
+def prod (s : Subsemigroup M) (t : Subsemigroup N) :
+    Subsemigroup (M × N) where
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:580]:
```suggestion
def prod (s : Subsemigroup M) (t : Subsemigroup N) : Subsemigroup (M × N) where
```

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 15]
```diff
@@ -0,0 +1,899 @@
...
+/-- Restriction of a semigroup hom to its range interpreted as a subsemigroup. -/
+@[to_additive "Restriction of an `AddSemigroup` hom to its range interpreted as a subsemigroup."]
+def srangeRestrict {N} [Mul N] (f : M →ₙ* N) : M →ₙ* f.srange :=
+  (f.codRestrict f.srange) fun x => ⟨x, rfl⟩
+#align mul_hom.srange_restrict MulHom.srangeRestrict
+
+@[simp, to_additive]
+theorem coe_srange_restrict {N} [Mul N] (f : M →ₙ* N) (x : M) : (f.srangeRestrict x : N) = f x :=
+  rfl
+#align mul_hom.coe_srange_restrict MulHom.coe_srange_restrict
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:764]:
```suggestion
theorem coe_srangeRestrict {N} [Mul N] (f : M →ₙ* N) (x : M) : (f.srangeRestrict x : N) = f x :=
  rfl
#align mul_hom.coe_srange_restrict MulHom.coe_srangeRestrict
```

Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:764]:
```suggestion
theorem coe_srangeRestrict {N} [Mul N] (f : M →ₙ* N) (x : M) : (f.srangeRestrict x : N) = f x :=
  rfl
#align mul_hom.coe_srange_restrict MulHom.coe_srangeRestrict
```

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean [hunk 16]
```diff
@@ -0,0 +1,899 @@
...
+
+@[to_additive]
+theorem srange_restrict_surjective (f : M →ₙ* N) : Function.Surjective f.srangeRestrict :=
+  fun ⟨_, ⟨x, rfl⟩⟩ => ⟨x, rfl⟩
+#align mul_hom.srange_restrict_surjective MulHom.srange_restrict_surjective
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subsemigroup/Operations.lean:769]:
```suggestion
theorem srangeRestrict_surjective (f : M →ₙ* N) : Function.Surjective f.srangeRestrict :=
  fun ⟨_, ⟨x, rfl⟩⟩ => ⟨x, rfl⟩
#align mul_hom.srange_restrict_surjective MulHom.srangeRestrict_surjective
```

**Review by Ruben-VandeVelde — DISMISSED** [review]

File: Mathlib/GroupTheory/Subsemigroup/Operations.lean (hunk 8 above)
Comment (ChrisHughes24) [reply Mathlib/GroupTheory/Subsemigroup/Operations.lean:547]: This is all done

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**